File: sort-list/sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        
        return mergeSort(head);   
    }
    
    public ListNode mergeSort(ListNode head){
        if(head == null || head.next == null){
            return head;
        }
        
        ListNode middle = getMiddle(head);
        ListNode nextOfMiddle = middle.next;
        middle.next = null;
        
        ListNode left = mergeSort(head);
        ListNode right = mergeSort(nextOfMiddle);
        return sortMerge(left, right);
        
    }
    
    public ListNode sortMerge(ListNode left, ListNode right){
        
        if(left == null){
            return right;
        }
        
        if(right == null){
            return left;
        }
        
        ListNode temp = null;
        
        if(left.val <= right.val){
            temp = left;
            temp.next = sortMerge(left.next, right);
        }
        else{
            temp = right;
            temp.next = sortMerge(left, right.next);
        }
        return temp;
    }
    
    public ListNode getMiddle(ListNode head){
        if(head == null){
            return head;
        }
        
        ListNode slow = head;
        ListNode fast = head;
        
        while(fast.next != null && fast.next.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }  
}

File: sort-list/sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if(head == NULL || head->next == NULL){
            return head;
        }
        ListNode* fast = head->next;
        ListNode* slow = head;
        
        while(fast != NULL && fast->next != NULL){
            slow = slow->next;
            fast = fast->next->next;
        }
        
        fast = slow->next;
        slow->next = NULL;
        return merge(sortList(head), sortList(fast));
    }
    
    ListNode * merge(ListNode* left, ListNode* right){
        ListNode* cur = new ListNode(0);
        ListNode* temp = cur;
        
        while(left != NULL && right != NULL){
            if(left->val < right->val){
                temp->next = left;
                temp = left;
                left = left->next;
            }
            else{
                temp->next = right;
                temp = right;
                right = right->next;
            }
        }
        if(left != NULL){
            temp->next = left;
        }
        else if(right != NULL){
            temp->next = right;
        }
        return cur->next;
    }
};